src/usr/local/www/interfaces_gre_edit.php

```diff
@@ -55,38 +59,80 @@
 	$pconfig['remote-addr'] = addrtolower($_POST['remote-addr']);
 
+	$tunnel_type = '';
+	if (($pconfig['tunnel-local-addr'] !== '') || ($pconfig['tunnel-remote-addr'] !== '')) {
+		$tunnel_type = 'v4';
+	}
+	if (($pconfig['tunnel-local-addr6'] !== '') || ($pconfig['tunnel-remote-addr6'] !== '')) {
+		$tunnel_type .= 'v6';
+	}
+
 	/* input validation */
-	$reqdfields = explode(" ", "if remote-addr tunnel-local-addr tunnel-remote-addr tunnel-remote-net");
-	$reqdfieldsn = array(gettext("Parent interface"), gettext("Remote tunnel endpoint IP address"), gettext("Local tunnel IP address"), gettext("Remote tunnel IP address"), gettext("Remote tunnel network"));
+	if ($tunnel_type === 'v4v6') {
+		$reqdfields = explode(" ", "if remote-addr tunnel-local-addr tunnel-remote-addr tunnel-remote-net tunnel-local-addr6 tunnel-remote-addr6 tunnel-remote-net6");
+		$reqdfieldsn = array(gettext("Parent interface"), gettext("Remote tunnel endpoint IPv4 address"), gettext("Local tunnel IPv4 address"), gettext("Remote tunnel IPv4 address"), gettext("Remote IPv4 tunnel network"), gettext("Remote tunnel endpoint IPv6 address"), gettext("Local tunnel IPv6 address"), gettext("Remote tunnel IPv6 address"), gettext("Remote IPv6 tunnel network"));
+	} else if ($tunnel_type === 'v6') {
+		$reqdfields = explode(" ", "if remote-addr tunnel-local-addr6 tunnel-remote-addr6 tunnel-remote-net6");
+		$reqdfieldsn = array(gettext("Parent interface"), gettext("Remote tunnel endpoint IPv6 address"), gettext("Local tunnel IPv6 address"), gettext("Remote tunnel IPv6 address"), gettext("Remote IPv6 tunnel network"));
+	} else {
+		$reqdfields = explode(" ", "if remote-addr tunnel-local-addr tunnel-remote-addr tunnel-remote-net");
+		$reqdfieldsn = array(gettext("Parent interface"), gettext("Remote tunnel endpoint IPv4 address"), gettext("Local tunnel IPv4 address"), gettext("Remote tunnel IPv4 address"), gettext("Remote IPv4 tunnel network"));
+	}
 
 	do_input_validation($_POST, $reqdfields, $reqdfieldsn, $input_errors);
 
-	if ((!is_ipaddr($_POST['tunnel-local-addr'])) ||
-	    (!is_ipaddr($_POST['tunnel-remote-addr'])) ||
-	    (!is_ipaddr($_POST['remote-addr']))) {
-		$input_errors[] = gettext("The tunnel local and tunnel remote fields must have valid IP addresses.");
+	if ($tunnel_type === '') {
+		$input_errors[] = gettext("The tunnel needs either a valid IPv4 or IPv6 tunnel configuration.");
+	}
+
+	if (((!is_ipaddr($_POST['tunnel-local-addr'])) ||
+	     (!is_ipaddr($_POST['tunnel-remote-addr']))) &&
+	    in_array($tunnel_type, array('v4', 'v4v6'))) {
+		$input_errors[] = gettext("The tunnel local and tunnel remote fields must have valid IPv4 addresses.");
+	}
+
+	if (((!is_ipaddrv6($_POST['tunnel-local-addr6'])) ||
+	     (!is_ipaddrv6($_POST['tunnel-remote-addr6']))) &&
+	     in_array($tunnel_type, array('v6', 'v4v6'))) {
+		$input_errors[] = gettext("The tunnel local and tunnel remote fields must have valid IPv6 addresses.");
```

Comment:
Since these are IPv6-specific the text should mention the exact labels of the IPv6 versions of these fields.